### ISSUE TITLE
[Feature] refreshKey() signature change - pass query context

### ISSUE THREAD

**Issue by dcworldwide** [cube-js/cube.js#183]
## tldr; 

We want to do something like `refreshKey: (tenantId) => SELECT MAX(updatedAt) FROM orders where tenantId=${tenantId}`

## Details

Requirement:

- We have time series tables with multiple tenants in the same table, isolated by a tenantId column 
- We want to rollup/pre-aggregate these tables into partitions (monthly). 
- We want to create a custom refresh key function that will calculate which partitions are to be invalidated and re-created (as per https://cube.dev/docs/incremental-pre-aggregations). 

Issue:

- Because our time series tables have multi-tenant data within the same table, we want to use a tenantId column to isolate the data. This allows us to do something like `refreshKey: (tenantId) => SELECT MAX(updatedAt) FROM orders where tenantId=${tenantId}`
- Our userContext object is a JWT that contains 1.* tenant Ids (a user has multiple tenants in our system). So we need a way to know which tenantId the query is for

Assumptions:

- I've assumed that our single-table, multi-tenant architecture is acceptable design. Open to opinions about this.

Love the lib so far, we are evaluating using cubejs instead of our custom analytics engine. 

Warm regards

**Comment by paveltiunov** [comment]
Hey @dcworldwide ! It's a great question! We just recently discussed it in Slack: https://cube-js.slack.com/archives/CC0403RRR/p1565487977139700?thread_ts=1565461573.136600&cid=CC0403RRR.

If you want to combine row level security together with pre-aggregations `queryTransformer` is a better fit for you then: https://cube.dev/docs/@cubejs-backend-server-core#options-reference-query-transformer.

**Comment by dcworldwide** [comment]
Hi @paveltiunov 

Sorry, I'm not sure I understand how this addresses my requirement... I understand that I can use queryTransformer to mutate the query and implicitly set a tenantId filter. This works well for row level security. 

But my requirement is to customise the behaviour of refreshKey...I want to progamatically tell cube.js which partitions of a cube's pre-aggregations to re-create.  

I'm probably using the wrong terminology, so here is an example of what I'm trying to do:

- I have multiple tenants in a single time series table and that table is pre-aggregated into incremental partitions - A, B, C, D
-Each tenant's data will be split across each partition
-Each tenant's data changes are not 'append only'....sometimes, older rows of data are updated. 
-If only tenant X's data has changed and the last changed record for tenant X is in partition C, we want to re-create partition C and D only.  That is why we need to pass a tenantId into refreshKey() so that we can calculate the last updated record's date for a specific tenant.

**Comment by paveltiunov** [comment]
@dcworldwide I see. What's your final goal you want to achieve by implementing this behavior? Do you want to reduce unnecessary pre-aggregation refreshes?

**Comment by dcworldwide** [comment]
Yes that's right, we want fine grained control over which monthly partitions of a pre-aggregation are re-built.

**Comment by paveltiunov** [comment]
@dcworldwide Makes sense. What database do you use and how many rows per tenant and in total? Is data partitioned by tenant id? 

There're usually two scenarios:
1. Immutable data partitioned by time. This way partitions built incrementally and all tenants data prepared at once at fixed cost of time partition preparation. Mutable data can't benefit a lot from time partitioning here because you'd need to refresh all the partitions all the time.
2. Unevenly distributed data across tenants that partitioned by tenant id. In this scenario you should go with multitenancy scenario where each tenant is isolated. You end up with per tenant pre-aggregations here.

You still can use `USER_CONTEXT` in `refreshKey` but it won't lead to desired result as every tenant touch will lead to different `refreshKey` which will in fact rebuild pre-aggregation for each query. Seems like you're asking here more about if-modified-since feature which should have a little bit different behavior.

**Comment by dcworldwide** [comment]
Thanks for the feedback.

We currently have 2,500 tenants, but that number will easily grow to 10/15k over time. Each tenant can have anywhere between 10k and 1m rows.

We prefer to go with a single table design, with the tenantId used as a dimension/filter. Because I believe this keeps things simple and manageable. 

How we partition this, I'm open to recommendations. If you think partitioning by tenantId is the optimal solution, we are interested in exploring it. I couldn't find any details on how it can be achieved via the docs however https://cube.dev/docs/pre-aggregations. Is it possible to partition your pre-aggregations by tenantId, using a single cube? I would not want to have to create 15k cubes, 1 per tenant for the same table.

**Comment by paveltiunov** [comment]
@dcworldwide Given the data is skew for different tenants I suggest you to go with per tenant partitioning through multitenancy: https://cube.dev/docs/multitenancy-setup. In this case you'll have only one cube defined for all tenants but each tenant will have own set of pre-aggregations.

**Comment by dcworldwide** [comment]
Hi @paveltiunov Could you please update that page to include an example that is specific to the use case I'm trying to achieve? single db, single cube with pre-aggs partitioned by tenantId.

I think many people will have the same question in the future as it is a simple and therefore common design.

**Comment by paveltiunov** [comment]
@dcworldwide Yep! Sure! Could you please check this doc first: https://cube.dev/docs/@cubejs-backend-server-core#options-reference-pre-aggregations-schema. I'll try to add this section to Multitenancy doc meanwhile.

**Comment by dcworldwide** [comment]
Awesome, that's great to hear :)

p.s. Please keep in mind that in our case, our user's JWT (authInfo) has 1.* tenantIds. I.e. each user, has multiple tenants. But, will only ever query data from 1 tenant at any one time. 

Example of our user JWT

```
{
  "sub": "1234567890",
  "name": "John Doe",
  "tenants": [
    {
      "id": 1,
      "role": "admin"
    },
    {
      "id": 2,
      "role": "admin"
    },
    {
      "id": 3,
      "role": "user"
    }
  ]
}
```

**Comment by dcworldwide** [comment]
As per my last comment, does the `preAggregationsSchema` signature support the following? That would provide an ideal solution. 

```
CubejsServerCore.create({
  preAggregationsSchema: ({ query }) => `pre_aggregations_${query.tenantId}`
});
```

With this method, is the cube name automatically added to the `pre_aggregations_${query.tenantId}` string?

Best,

**Comment by paveltiunov** [comment]
@dcworldwide Pre-aggregations schema name is per application id as it cached on this level. Just added example here: https://cube.dev/docs/multitenancy-setup#same-db-instance-with-per-tenant-pre-aggregations.

So using `query` inside `preAggregationsSchema` isn't supported right now. What you can do is to provide current `tenantId` as part of token for each query which will select right app based on `contextToAppId`. Another solution is to use concatenation of `tenants` in `contextToAppId`. This way it'll create new app and pre-aggregations schema for each combination of tenants.

**Comment by dcworldwide** [comment]
@paveltiunov Nice work on the doc update. Already much clearer to read. Hopefully that means less questions for you in the future :)

> So using query inside preAggregationsSchema isn't supported right now. What you can do is to provide current tenantId as part of token for each query which will select right app based on contextToAppId.

Ok, that is workable for us. The original motivation for my question - is that because we already have a REST/GQL api with a JW, we were hoping to re-use that JWT with cubejs to eliminate the need to call JWT.sign() for each cubejs request. This is simply a performance consideration.

> All pre-aggregations stored as a separate tables in pre-aggregations schema

That is a good design.

**Comment by paveltiunov** [comment]
@dcworldwide Thanks! Yep. You still can re-use token if you implement your custom `checkAuthMiddleware`: https://cube.dev/docs/@cubejs-backend-server-core#options-reference-check-auth-middleware. This way you can update `authInfo` accordingly to your needs. Please note that in order to use `USER_CONTEXT` in cube definition you'd need to pass `u` property in your JWT token: https://cube.dev/docs/security#security-context.

**Comment by dcworldwide** [comment]
Morning @paveltiunov , I think we can close this for now. I will re-open if our POC implementation has issues. Please re-open if I've closed this prematurely. Thanks for your assistance.
